From: Pekka Helenius <fincer89@hotmail.com>
Date: Tue, 04 Aug 2020 01:52:13 +0300
Subject: Fix debug mode not showing output in command line


--- a/src/ntpd.c	2020-08-03 23:30:09.431955200 +0300
+++ b/src/ntpd.c	2020-08-03 23:29:45.150837701 +0300
@@ -195,12 +195,14 @@ main(int argc, char *argv[])
 		switch (ch) {
 		case 'd':
 			lconf.debug = 1;
+			lconf.verbose = 2;
 			break;
 		case 'f':
 			conffile = optarg;
 			break;
 		case 'n':
 			lconf.debug = 1;
+			lconf.verbose = 2;
 			lconf.noaction = 1;
 			break;
 		case 'P':
